includes/Checker/Abstract_Check_Runner.php

```diff
@@ -0,0 +1,179 @@
+<?php
+/**
+ * Class WordPress\Plugin_Check\Checker\Abstract_Check_runner
+ *
+ * @package plugin-check
+ */
+
+namespace WordPress\Plugin_Check\Checker;
+
+use WordPress\Plugin_Check\Checker\Check_Runner;
+use WordPress\Plugin_Check\Checker\Check_Result;
+use WordPress\Plugin_Check\Checker\Preparations\Universal_Runtime_Preparation;
+
+/**
+ * Abstract Check Runner class.
+ *
+ * @since n.e.x.t
+ */
+abstract class Abstract_Check_Runner implements Check_Runner {
+
+	/**
+	 * Instance of the Checks class.
+	 *
+	 * @since n.e.x.t
+	 * @var Checks
+	 */
+	protected $checks;
+
+	/**
+	 * Instance of the Check_Context class.
+	 *
+	 * @since n.e.x.t
+	 * @var Check_Context
+	 */
+	protected $context;
+
+	/**
+	 * Array of Check instances to run.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $checks_to_run;
+
+	/**
+	 * Determines if the current request is intended for the plugin checker.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return boolean Returns true if the check is for plugin else false.
+	 */
+	abstract public function is_plugin_check();
+
+	/**
+	 * Setup the check context, checks and preparations based on the request.
+	 *
+	 * @since n.e.x.t
+	 */
+	abstract protected function setup_checks();
+
+	/**
+	 * Prepare the environment for running the requested checks.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Cleanup function to revert any changes made here.
+	 *
+	 * @throws Exception Thrown exception when preparation fails.
+	 */
+	public function prepare() {
+		if ( ! $this->requires_universal_preparations( $this->checks_to_run ) ) {
+			$preparation = new Universal_Runtime_Preparation( $this->context );
+			return $preparation->prepare();
+		}
+
+		return null;
+	}
+
+	/**
+	 * Run the checks against the plugin.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return Check_Result
+	 */
+	public function run() {
+		$checks = $this->checks->get_checks( $this->checks_to_run );
```

Comment:
Related to the above, `$this->checks` will be undefined if `setup_checks()` is not called, or implemented incorrectly.

Seeing this, we should probably avoid the `$this->checks` and `$this->checks_to_run` class properties entirely. It would be a much clearer API to have e.g. `setup_checks()` _return_ the `Checks` instance rather than implicitly setting it in a class property, as that would then be clarified already by the method documentation.

Not sure how to best handle `$this->checks_to_run` - maybe introduce another abstract protected method like `get_checks_to_run()` that returns that list of checks to run?